providers/r2dbc/shedlock-provider-r2dbc/src/main/java/net/javacrumbs/shedlock/provider/r2dbc/R2dbcStorageAccessor.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2009 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package net.javacrumbs.shedlock.provider.r2dbc;
+
+import io.r2dbc.spi.Connection;
+import io.r2dbc.spi.ConnectionFactory;
+import io.r2dbc.spi.Statement;
+import net.javacrumbs.shedlock.provider.r2dbc.internal.AbstractR2dbcStorageAccessor;
+import net.javacrumbs.shedlock.support.annotation.NonNull;
+import reactor.core.publisher.Mono;
+
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import static java.util.Objects.requireNonNull;
+
+class R2dbcStorageAccessor extends AbstractR2dbcStorageAccessor {
+    private final ConnectionFactory connectionFactory;
+
+    R2dbcStorageAccessor(@NonNull ConnectionFactory connectionFactory, @NonNull String tableName) {
+        super(tableName);
+        this.connectionFactory = requireNonNull(connectionFactory, "dataSource can not be null");
+    }
+
+    @Override
+    protected <T> Mono<T> executeCommand(
+        String sql,
+        Function<Statement, Mono<T>> body,
+        BiFunction<String, Throwable, Mono<T>> exceptionHandler
+    ) {
+        return Mono.usingWhen(
```

Comment:
Wow, that's why I really hope that Project Loom will be a thing. How is anyone supposed to understand code like this? Just ranting, it's not your fault.

Reply:
It's a matter of habit actually. 
And yes, in some cases reactive and FP libraries reinvent a wheel with such constructions (in this case try-with-resources).